src/Core/Block.cpp

```diff
@@ -139,15 +139,17 @@ void Block::insert(size_t position, ColumnWithTypeAndName elem)
     if (elem.name.empty())
         throw Exception("Column name in Block cannot be empty", ErrorCodes::AMBIGUOUS_COLUMN_NAME);
 
-    for (auto & name_pos : index_by_name)
-        if (name_pos.second >= position)
-            ++name_pos.second;
-
-    auto [it, inserted] = index_by_name.emplace(elem.name, position);
+    auto [it, inserted] = index_by_name.emplace(elem.name, std::npos);
```

Comment:
Better to not use emplace with std::npos, just find element, if it already exists then checkColumnStructure, after that we can safely update index_by_name and insert element.

Reply:
Yes, better way. I will refactor it as you said.

Reply:
The current code changes are supposed to comply with the original code semantically and can solve the potential inconsistent state in Block. However, I need to point out, for the case of  columns with both identical name and identical structure, the new index_by_name entry still has NOT been established for conflicting name. This is tolerable because both old index_by_name entry and old data entry still exist. It has no other side effects except one new data entry being inserted with no index pointing to it. It is somewhat complicated to completely get rid of old index_by_name entry and old data and then replace them with the corresponding new ones. And I also don't know if this logic is the right one you want.